### ISSUE TITLE
Programmatic definition of data sets

### ISSUE THREAD

**Issue by raderio** [database-rider/database-rider#129]
Is it possible to programmatic define data sets?

Something like in jdbdt http://jdbdt.org/DataSets.html
```
DataSet data = 
   builder(t)
  .sequence("ID", 1) // 1, 2, 3, ...
  .sequence("LOGIN", "harry", "mark", "john")
  .sequence("NAME", "Harry H", "Mark M", "John J")
  .sequence("PASSWORD", i -> "password " + i , 1)
  .random("CREATED", Date.valueOf("2015-01-01"), Date.valueOf("2015-12-31"))
  .generate(3) // generate 3 rows 
  .sequence("LOGIN", i -> "guest_" + i, 4)  // "user_4", "user_5", ...
  .sequence("NAME", i -> "Guest User " + i, 4) // "Guest User 4", ...
  .value("password", "samePasswordForAllGuests") 
  .generate(6) // 6 more rows keeping ID sequence and CREATED random filler
  .data();   
```

This will improve refactoring in case of database schema changes, also generation can be simplified.

**Comment by rmpestano** [comment]
Hi, currently you can only generate dataset files from database tables via @ExportDataSet, via dataset exporter API or using dbunit addon, see here: https://github.com/database-rider/database-rider#export-datasets.

I think it would be nice to have something like jdbdt which generates dbunit datasets

**Comment by raderio** [comment]
Maybe this can be implemented in functional way, not with annotations, something like
```
@Test
public void shouldLoadUserFollowers() {
    dbSetup(Global.usersDataSet)
    // ...
}
```

**Comment by raderio** [comment]
Or it could be implemented as in DbSetup, from Ninja Squad http://dbsetup.ninja-squad.com/user-guide.html
```
public class VendorRepositoryTest {

    private DataSource dataSource = ...;

    @Before
    public void prepare() throws Exception {
        Operation operation =
            sequenceOf(
                CommonOperations.DELETE_ALL,
                CommonOperations.INSERT_REFERENCE_DATA,
                insertInto("VENDOR")
                    .columns("ID", "CODE", "NAME", "COUNTRY_ID")
                    .values(1L, "AMA", "Amazon", 2)
                    .values(2L, "PMI", "PriceMinister", 1)
                    .build());
        DbSetup dbSetup = new DbSetup(new DataSourceDestination(dataSource), operation);
        // or without DataSource:
        // DbSetup dbSetup = new DbSetup(new DriverManagerDestination(TEST_DB_URL, TEST_DB_USER, TEST_DB_PASSWORD), operation);
        dbSetup.launch();
    }
}
```

**Comment by rmpestano** [comment]
What i'm thinking is to create an api which would generate dataset files so they can be used in tests. I don't like to mix infrastructure code like populate database with the test itself. 

In you example above why would you need dbrider at all? I may be missing something but I think dbsetup already does what you need, don't It?

**Comment by raderio** [comment]
Dbsetup do not have expect database functionality.

Regarding file generation, maybe in  `@DataSet` annotation we can specify a `.class` where will be a method that will return `DataSet` object, so no need for file generation?

**Comment by raderio** [comment]
```
class UsersDataSetProvider implements DataSetProvider {
    public DataSet provide() {
        DataSet data = 
   builder(t)
  .sequence("ID", 1) // 1, 2, 3, ...
  .sequence("LOGIN", "harry", "mark", "john")
  .sequence("NAME", "Harry H", "Mark M", "John J")
  .sequence("PASSWORD", i -> "password " + i , 1)
  .random("CREATED", Date.valueOf("2015-01-01"), Date.valueOf("2015-12-31"))
  .generate(3) // generate 3 rows 
  .sequence("LOGIN", i -> "guest_" + i, 4)  // "user_4", "user_5", ...
  .sequence("NAME", i -> "Guest User " + i, 4) // "Guest User 4", ...
  .value("password", "samePasswordForAllGuests") 
  .generate(6) // 6 more rows keeping ID sequence and CREATED random filler
  .data();   
    return data;
    } 
}
```
```
@Test
@DataSet(provider = UsersDataSetProvider.class)
public void shouldLoadUserFollowers() {
    User user = (User) emProvider.em().createQuery("select u from User u left join fetch u.followers where u.id = 1").getSingleResult();
    assertThat(user).isNotNull();
    assertThat(user.getId()).isEqualTo(1);
    // ...
}
```

**Comment by rmpestano** [comment]
Yea I like it, also we would not need to add any extra dependency to dbrider because It is up to the user decide how the dataset will be generated (jooq, JPA, JDBC etc..) right?

**Comment by rmpestano** [comment]
Well in fact the `dataset builder` needed to be provided by dbrider i think..

**Comment by rmpestano** [comment]
Maybe we can reuse jdbdt api do generate the dataset, e.g:
```
     public DataSet provide() {
        return 
            DatasetBuilder.
                config(new DataSetConfig().cleanAfter(true).strategy(SeedStrategy.CLEAN_INSERT)).
                data(builder(t) //jdbdt dataset builder
                    .sequence("ID", 1) // 1, 2, 3, ...
                    .sequence("LOGIN", "harry", "mark", "john")
                    .sequence("NAME", "Harry H", "Mark M", "John J")
                    .sequence("PASSWORD", i -> "password " + i , 1)
                    .random("CREATED", Date.valueOf("2015-01-01"), Date.valueOf("2015-12-31"))
                    .generate(3) // generate 3 rows 
                    .sequence("LOGIN", i -> "guest_" + i, 4)  // "user_4", "user_5", ...
                    .sequence("NAME", i -> "Guest User " + i, 4) // "Guest User 4", ...
                    .value("password", "samePasswordForAllGuests") 
                    .generate(6) // 6 more rows keeping ID sequence and CREATED random filler
                    .data()). //use jdbdt dataset builder to create a dataset https://github.com/JDBDT/jdbdt/blob/master/src/main/java/org/jdbdt/DataSet.java
                .create(); //iterare over jdbdt dataset rows and generate dbunit dataset files to used in `value` property of DataSet https://github.com/database-rider/database-rider/blob/21d5043e08b18a594a659e2508014fa9fba77fab/rider-core/src/main/java/com/github/database/rider/core/api/dataset/DataSet.java#L18
                  
    }
``` 
I don't have experience with jdbdt so I don't know if it is possible to generate it's dataset without populating the database becase my idea is to generate a DataSet and pass it to [DatasetExecutor as we do today](https://github.com/database-rider/database-rider/blob/21d5043e08b18a594a659e2508014fa9fba77fab/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L109). Note that the DataSetConfig can be [created from DataSet](https://github.com/database-rider/database-rider/blob/21d5043e08b18a594a659e2508014fa9fba77fab/rider-core/src/main/java/com/github/database/rider/core/configuration/DataSetConfig.java#L118).

Also I don't know how it will work for dependent tables, e.g how to generate the following dataSet:

```
USER:
  - ID: 1
    NAME: "@realpestano"
  - ID: 2
    NAME: "@dbunit"
TWEET:
  - ID: abcdef12345
    CONTENT: "dbunit rules!"
    DATE: "[DAY,NOW]"
    USER_ID: 1
FOLLOWER:
  - ID: 1
    USER_ID: 1
    FOLLOWER_ID: 2
```

WDYT?

**Comment by rmpestano** [comment]
Guys, what do you think of such feature, do you think it would be useful?  

@xpinjection @iceberk @arhohuttunen @artemy-osipov  @getarrenspieler @njuro @jenskreidler

**Comment by xpinjection** [comment]
From my point of view datasets should be stored in more appropriate format than code. If you would like to insert data into DB from object model - use JPA. 

To support refactoring of the database it is much better to use XML and continuously generate XSD/DTD for validation and schema compatibility control.

**Comment by artemy-osipov** [comment]
If we only considering the possibility of building datasets programmatically, I think it's a much more powerful and flexible way. Datasets and any other common test data used in setup and verify phases are very similar in nature and it is understandable to keep them in the same format and location. In either case, the datasets will eventually be converted into some structure in memory to load them into the database, whether it is read from XML or provided from test code.

**Comment by xpinjection** [comment]
It is already possible to build dataset from the code, but stored in external form. :)

**Comment by arhohuttunen** [comment]
In my view, the whole reason of existence for dbunit or database-rider is that you don’t need to setup your test data in code format.

**Comment by artemy-osipov** [comment]
I would separate the concept of setup/verify state of the database and describing test data.

> It is already possible to build dataset from the code, but stored in external format. :)

Do you mean programmatically create xml in fs and specify it in DataSet?

**Comment by rmpestano** [comment]
Thank you very much for your time (on a weekend) and valuable opinion!

> If we only considering the possibility of building datasets programmatically, I think it's a much more powerful and flexible way

I agree, all the dynamism we wanted in datasets via groovy/javascript in datasets or regex in expected datasets are granted for free on the programatic approach.

> From my point of view datasets should be stored in more appropriate format than code. If you would like to insert data into DB from object model - use JPA.

In fact in the code approach we would not store the dataset but would generate it, and save in a temp folder for example, every time the test runs (if we implement such approach we should consider some sort of cache to not generate datasets unnecessarily) 

> In my view, the whole reason of existence for dbunit or database-rider is that you don’t need to setup your test data in code format.  

Yes, I agree but the programmatic approach brings another use case which is: If your schema is updated then your datasets based on code would be also updated more easily. 

Ideally if the dataset builder could be based on a `type safe solution` i.e [JPA metamodel ](https://thoughts-on-java.org/static-metamodel/) then the datasets based on code would be updated automatically, at least column/table rename so I think that if we agree on such approach we should aim for that.

WDYT?

**Comment by raderio** [comment]
> I don't have experience with jdbdt so I don't know if it is possible to generate it's dataset without populating the database

jdbdt is just an example of how to populate database programmatic. So, database-rider can provide own way, maybe a simple Map<String, Object>, or another solution.

Basically is the same idea like with files, but instead to generate that data programmatically, this way is more flexible and powerful, we can use cycles, variables, column refactoring for free, etc.

**Comment by arhohuttunen** [comment]
https://github.com/opensource21/dbunit-datasetbuilder

**Comment by raderio** [comment]
Yeah, something like that.
Is that repository maintained?

**Comment by rmpestano** [comment]
> https://github.com/opensource21/dbunit-datasetbuilder

Interesting! The nice thing is that we can use dbrider [yaml and json](https://github.com/database-rider/database-rider/tree/master/rider-core/src/main/java/com/github/database/rider/core/dataset/writer) writer on top of that.

I would replace the `ColumnSpec` with JPA metamodel, somethink like:

DataSetBuilder builder = new DataSetBuilder();
builder.newRow("PERSON").
        with(Person_.id, 1L). 
        with(Person_.name, "Bob").
        with(Person_.age, 18).
        with(Person_.date, "[DAY,NOW]").
        with(Address_.city, "Porto").
      add();


File datasetFile = Files.createTempFile(fileName, ".yml").toFile();
FileOutputStream fos = new FileOutputStream(datasetFile);
new YMLWriter(fos).write(builder.build());
or 
new JSONWriter(fos,dataSet).write();

DataSet dataSet = new DataSetImpl(datasetFile.toAbsolutePath(), executorId(), strategy(), useSequenceFiltering(), tableOrdering(), disableConstraints(), fillIdentityColumns(), executeStatementsBefore, executeScriptsAfter, executeScriptsBefore, executeStatementsAfter, cleanBefore(), cleanAfter(), transactional());

Anyone willing to implement?

**Comment by rmpestano** [comment]
Guys, I've just opened the PR implementing this feature,

Any feedback is really welcome, thank you all for all the insights so far!

**Comment by raderio** [comment]
Please add also possibility to store some rows in a variable and reuse it in different datasets, so `DataSetBuilder` can be build by providing some predefined rows.

**Comment by rmpestano** [comment]
Hi @raderio, good idea, just added support for it, [see an example here](https://github.com/database-rider/database-rider/blob/35cc5e78f92f824d0b395caa7e15983fead35edb/rider-core/src/test/java/com/github/database/rider/core/DataSetProviderIt.java#L222-L237). 

We are still discussing the builder API, see here: https://github.com/database-rider/database-rider/pull/132#discussion_r283108511

What do you think would be more clear:

```
builder.row("TWEET")
                .column("ID", "abcdef12345").column("CONTENT", "dbrider rules!")
                .column("DATE", "[DAY,NOW]").add();
```

or 

```
builder.newRow("TWEET")
                .with("ID", "abcdef12345").with("CONTENT", "dbrider rules!")
                .withDATE", "[DAY,NOW]").add();
```

or other approach?

**Comment by rmpestano** [comment]
I think the more cleaner approach would be:
```
builder.rowForTable("TWEET")
                .withColumn("ID", "abcdef12345").withColumn("CONTENT", "dbrider rules!")
                .withColumn("DATE", "[DAY,NOW]").add();
```

but I think it is too verbose so that's why I prefer the first approach which I read as:

```
builder.row("TWEET")//start the creation of a row for table tweet
             .column("ID", "abcdef12345") //adds a column on current row
             .column("CONTENT", "dbrider rules!") //another column on current row
.           add() //add the row to the dataset being build
           .row("TWEET") //starts another row for the same table (could be a different table) and so on
           .build()//creates the dataset with added rows
```

**Comment by raderio** [comment]
I like more row/column approach.

BTW, maybe we do not need to call `add()`, `row` and `build` can be a delimiter, what do you think about
```
builder.row("TWEET")
           .column("ID", "abcdef12345").column("CONTENT", "dbrider rules!")
           .column("DATE", "[DAY,NOW]")
       .row("PERSON").
           .column("ID", 1L). 
           .column("NAME", "Bob")
       .build();
```

**Comment by rmpestano** [comment]
Hi @raderio, It would be nice but the problem is that we need to add the row to the current dataset being created only after it is finished (I think), see here: https://github.com/rmpestano/database-rider/blob/1d4d1dfc270d33c2f7e28fb0018d94997db014d1/rider-core/src/main/java/com/github/database/rider/core/dataset/builder/DataRowBuilder.java#L80

**Comment by raderio** [comment]
We can call
```
dataSet.add(this);
```
inside `row` method.

However, I am also ok with `add` method.

**Comment by getarrenspieler** [comment]
@rmpestano ,

> ```java
>      builder.rowForTable("TWEET")
>             .withColumn("ID", "abcdef12345").withColumn("CONTENT", "dbrider rules!")
>             .withColumn("DATE", "[DAY,NOW]").add();
>  ```

What if table mention will be more strict with `table(String name)` method? It can be used for any shortcuts for the table values:
```java
builder.table("TWEET") //table definition is used until next table mention in table() or row(String) method
     .default("date", "[NOW]") // values as a rule used the same manner on all rows
     .default("deleted", "[null]") // and putting same placeholders/values on each row is tedious 
     .row() 
     .column("id", "id1")
     .column("another", 5)
     ....
     .row()    // adding row to the table TWEET
     .column("id", "id2")
     ...
     .row("NEW_TABLE") //syntax for adding single row is still valid for short datasets
     .append("key", 6)
     ....
     .build();
```

Builder itself can use `default` methods to produce values for columns for all rows in all tables. I thing that is a common when columns with the same usage are named identically in different tables.

**Comment by getarrenspieler** [comment]
As for column/withColumn etc for column values setting method naming I think it should be as short as possible.

**Comment by rmpestano** [comment]
Hi @jenskreidler, 
 
> Builder itself can use default methods to produce values for columns for all rows in all tables. I thing that is a common when columns with the same usage are named identically in different tables.

Yes I saw it on [dbunit-datasetbuilder](https://github.com/opensource21/dbunit-datasetbuilder) but removed to simplify things but now as I got better understanding of the builder I think it is easier to add back.

`builder.table("TWEET") //table definition is used until next table mention in table() or row(String) method`

I'm not sure, I think it would cause confusion, I'm inclined to stay with the  row("TABLE") approach, I think with good documentation and examples we can overcome it's lack of readability, WDYT?

**Comment by getarrenspieler** [comment]
@rmpestano, 

> I think it would cause confusion

Yes there is some risk.  I mean there needed any simple syntax to work with numerous rows because paste table table name each time per row is may be too verbose (in spite of it can be set with constant).

jdbt approach in this case (as stated earlier) is looks like cumbersome as you should follow order of  values in per row tuples (may be there is another way, my opinion is about examples in posts here).

When you have to build dataset with more than 10 rows with about 12 columns it's rather hard to count and read values written this way.

Current table fixation can decrease verbocity a little. But may be it should be any special builder for "batch building".

P.S. I like row(String table) syntax though)

**Comment by raderio** [comment]
As first version I think is ok `row(tableName).column`.

Another builder can be added later. For example
```
    builder.row(tableName).column // first builder
    builder.table(tableName).row().column // another builder
```

**Comment by rmpestano** [comment]
Hi @raderio,

> We can call
> `dataSet.add(this);`
> inside row method.

Just refectory the builder to do that, now the add method is not necessary anymore, it is called internally. 

@jenskreidler,

Added the defaultValue support, [see example here](https://github.com/rmpestano/database-rider/blob/cbc2e82a47089be102b060ca04ad3b68a757c087/rider-core/src/test/java/com/github/database/rider/core/dataset/builder/DatasetBuilderTest.java#L175-L215)  


About the builder.table("") I still prefer the simple version using row("table") maybe later we can extend the builder to add it

Thank you very much for your insights

**Comment by rmpestano** [comment]
Well as now have a variable for storing current row being created, [see here](https://github.com/rmpestano/database-rider/blob/cbc2e82a47089be102b060ca04ad3b68a757c087/rider-core/src/main/java/com/github/database/rider/core/dataset/builder/DataSetBuilder.java#L58), It is quite easy to add a `row()` without passing the table and just reuse current row in this case. I'll add both table("tableName") and row() methods.

**Comment by rmpestano** [comment]
Here we go, in the end I replaced the `row("tableName")` with `table("tableName"), see example below:

```
builder
        .table("USER") //start adding rows to 'USER' table
            .column("ID",1)
            .column(name,"@realpestano")
        .row() //keeps adding rows to the current table
            .column(id,2)
            .column("NAME","@dbunit")
        .table("TWEET") //starts adding rows to 'TWEET' table
            .column("ID","abcdef12345")
            .column("CONTENT","dbunit rules!")
            .column("DATE","[DAY,NOW]")
        .table("FOLLOWER").column(id,1)
            .column("USER_ID",1)
            .column("FOLLOWER_ID",2)
        .table("USER")// we still can add rows to table already added to the dataset
            .column("ID",3)
            .column(name,"@new row")
        .build();

```

which generates the following dataset (in yaml):

```
FOLLOWER:
  - ID: 1
    USER_ID: 1
    FOLLOWER_ID: 2

TWEET:
  - ID: "abcdef12345"
    CONTENT: "dbunit rules!"
    DATE: "2019-05-14 19:07:22.0"

USER:
  - ID: 1
    NAME: "@realpestano"
  - ID: 2
    NAME: "@dbunit"
  - ID: 3
    NAME: "@new row"
```

What do you think?

**Comment by xpinjection** [comment]
I think with such syntax it is better to make row() call mandatory to avoid confusion.

**Comment by rmpestano** [comment]
Hi @xpinjection,

> I think with such syntax it is better to make row() call mandatory to avoid confusion.

Yes it is, you can't call row() directly in DataSetBuilder, you must "start a table" before, hence the row() method is only declared in [DataRowBuilder](https://github.com/rmpestano/database-rider/blob/d07063822ef620b2baf8e2aa16e0ae6f1b32efc3/rider-core/src/main/java/com/github/database/rider/core/dataset/builder/DataRowBuilder.java#L93), and table method is on the [DataSetBuilder](https://github.com/rmpestano/database-rider/blob/d07063822ef620b2baf8e2aa16e0ae6f1b32efc3/rider-core/src/main/java/com/github/database/rider/core/dataset/builder/DataSetBuilder.java#L61)

I don't know if you were talking about it...

**Comment by rmpestano** [comment]
Hmm I think I got what you mean:

```
builder
        .table("USER") 
            .row()
            .column("ID",1)
```

Is also possible but I don't thing it should be mandatory

**Comment by arhohuttunen** [comment]
I agree with @xpinjection that with such syntax it should be mandatroy to avoid confusion. I also think such syntax is more descriptive in relation to what is actually happening.

**Comment by raderio** [comment]
I also agree that `row` should be mandatory.

**Comment by rmpestano** [comment]
Majority wins, anyone else?

One thing that having a table("").row() would enable is to have defaultValues per table because we will need to add another class (TableBuilder) to the builder chain and it could hold default values for each table, of cource defaultValues for all tables would also be allowed, something like:

```
builder.defaultValue("id", -1).//defaultValue for all tables, applied only if not specified
              table("user")
                  .defaultValue("date", "[DAY, NOW]") //defaultValue only for table user
              .row()
                    .column("id", 1)
                    .column("name", "@realpestano")
              .row() //creates a new row for user table
                    .column("id", 2)
                    .column("name", "@dbunit")
              .row()
                   .column("name", "@dbunit3")//third row for user table using default value
              .table("tweet")
              .row()
                 . column("id", "abcdef12345").column("content", "dbunit rules!")
                  .column("date", "[DAY,NOW]")
              .table("follower")
              .row()
                  .column("id", 1)
                  .column("user_id", 1).column("follower_id", 2)
              .build();
```

**Comment by rmpestano** [comment]
Done, see changes here https://github.com/rmpestano/database-rider/commit/758b574c0719162a9b6b62b82d9e2e3d32b16270

**Comment by rmpestano** [comment]
Guys, what about the columns/values syntax, should we also support it?

```
builder  
              .table("user")
                  .columns("id", "name") 
                  .values(1,  "@dbrider")
                  .values(2,  "@dbunit")            
                  .values(3,  "@dbunit3")      
              .table("tweet")
                  .columns("id", "content", "date") 
                  .values("abcdef12345",  "dbunit rules!", "[DAY,NOW]")
              .table("follower")
                  .columns("id", "user_id", "follower_id") 
                  .values(1,  1, 2)
              .build();
``` 

In current syntax It would be the same as:

```
builder 
              .table("user")
              .row()
                    .column("id", 1)
                    .column("name", "@realpestano")
              .row() //creates a new row for user table
                    .column("id", 2)
                    .column("name", "@dbunit")
              .row()
                   .column("id", 3)
                   .column("name", "@dbunit3")
              .table("tweet")
              .row()
                  .column("id", "abcdef12345")
                  .column("content", "dbunit rules!")
                  .column("date", "[DAY,NOW]")
              .table("follower")
              .row()
                  .column("id", 1)
                  .column("user_id", 1)
                  .column("follower_id", 2)
              .build();
```
 
WDYT?

**Comment by xpinjection** [comment]
Asked several people, all of them voted for the second option.

**Comment by getarrenspieler** [comment]
@rmpestano 

> Guys, what about the columns/values syntax, should we also support it?

I think it's appropriate for small datasets with small column count.

row()/column() approach is cleaner

**Comment by rmpestano** [comment]
You think supporting both approaches would cause confusion?

**Comment by raderio** [comment]
I think columns/values can be added letter, as another feature, sometimes it can look nicer.

**Comment by xpinjection** [comment]
I think both approaches could cause confusion.

**Comment by rmpestano** [comment]
> I think columns/values can be added letter, as another feature, sometimes it can look nicer.

I already implemented it, we just need to decide if is is a good idea or not. 

I think in our daily work we deal with much more complex APIs, for example in JPA there is a plethora of ways to create a query (named, criteria, native, entityGraphs to name a few) the user just needs to decide which approach is better for his use case.

I am +1 for having both approaches

**Comment by arhohuttunen** [comment]
Actually this is related to what I suggested earlier:

```
builder 
              .table("user")
              .row()
                    .column("id").value(1)
                    .column("name").value("@realpestano")
              .build();
```

would be much more similar to:

```
builder  
              .table("user")
                  .columns("id", "name") 
                  .values(1,  "@dbrider")
              .build();
```

But having the both approaches in it's current form would cause confusion IMO.

**Comment by rmpestano** [comment]
> I don’t think it’s a fair comparison to named queries, criteria, etc. because those are different APIs but we are talking about a single API here.

agreed, here's  entityManager API:
![Screenshot from 2019-05-15 12-53-26](https://user-images.githubusercontent.com/1592273/57790082-735a7180-7710-11e9-9617-9a4b3872d5bb.png)

> If there would be both approaches then I think the latter one should be called row().columnWithValue(”id”, 1) or something similar because columns(”id”).values(1) and column(”id”, 1) would mean semantically really different things.

```columns(”id”).values(1)``` and ```column(”id”, 1)``` are excatly the same, will produce the same resuit

> But in all honesty, I don’t see myself using the columns().values() form, it seems less readable, mostly because varying parameter lengths.

Maybe it is a matter of taste but for small datasets (with lots of rows) I think the columns... values approach is much better, ex:

```
builder 
              .table("user")
              .row()
                    .column("id", 1)
                    .column("name", "@realpestano")
              .row() //creates a new row for user table
                    .column("id", 2)
                    .column("name", "@dbunit")
              .row()
                   .column("id", 3)
                   .column("name", "@dbunit3")
              .row()
                   .column("id", 4)
                   .column("name", "@dbunit4")
```

Is just:

```
builder  
              .table("user")
                  .columns("id", "name") 
                  .values(1,  "@dbrider")
                  .values(2,  "@dbunit")            
                  .values(3,  "@dbunit3")      
                  .values(4,  "@dbunit4")  
```

**Comment by arhohuttunen** [comment]
> columns(”id”).values(1) and column(”id”, 1) are exactly the same, will produce the same result

Yes, of course :) Let me try to rephrase what I meant: I think calling `columns()` to create columns and `values()` to create values is semantically different from calling `column()` to create both a column and it's value.

**Comment by rmpestano** [comment]
> Yes, for a simple case that is true. Depends how many columns there are. If there are many columns it becomes quickly less readable.

Sure, that's what I think also so its why I think we should have another option, here is the options an user of our API may have after starting a table:

![Screenshot from 2019-05-15 14-13-13](https://user-images.githubusercontent.com/1592273/57795315-09e06000-771c-11e9-99bf-63b5b779f467.png)

I don't think it can cause much confusion...

On the other hand, in order to reuse rows, we have an `add` directly on DataSetBuilder which  receives an already created row, this I think may cause confusion because it is the first contact with our API: 

![Screenshot from 2019-05-15 14-19-23](https://user-images.githubusercontent.com/1592273/57795588-bb7f9100-771c-11e9-8e94-ca404096fd18.png)

I thought in adding it to the tableBuilder but it doesn't make sense because the row was already created and already belongs to table...

**Comment by rmpestano** [comment]
Also an unintended side effect is that both approaches can be used together, this can be good if your DataSet has both simple tables with few columns (and many rows) and more complex tables where the first approach may fit better :

```
builder.defaultValue("id", -1)
                .table("user")
                    .columns("id", "name")
                    .values(1,  "@dbrider")
                    .values(2,  "@dbunit")
                    .values(null,  "@dbunit3")//will use default value
                .table("tweet")
                    .defaultValue("likes", 99)
                .row()
                        .column("id", "abcdef12345")
                        .column("content", "dbunit rules!")
                        .column("date", "[DAY,NOW]")
                .table("follower")
                    .columns("id", "user_id", "follower_id")
                    .values(1,  1, 2)
                .build();
```

**Comment by arhohuttunen** [comment]
Yeah I think it’s fine during development because of auto-completion. Mixing the two approaches is a perfect example of making it confusing readability wise though 😄 But I suppose the responsibility for that is left to the developer.

**Comment by rmpestano** [comment]
Did I messed everything? hahahahah

**Comment by arhohuttunen** [comment]
It’s sometimes easy to start overthinking things, so I think it’s actually pretty good.

**Comment by rmpestano** [comment]
Talking seriously, there is no much things one can do with such API, I have played a bit with some real datasets and one hardly will add a row to the wrong table or forget to add a value to a column because it is quite simple in respect to number of methods one can call.

Maybe I'm too used to it already that I can also easily read a dataset which uses the `mixed approach`

**Comment by raderio** [comment]
One use case for columns/values is when we want to add values in a cycle, so before cycle we define columns, and in cycle we set the values.

**Comment by rmpestano** [comment]
> One use case for columns/values is when we want to add values in a cycle, so before cycle we define columns, and in cycle we set the values.

Yes, I'm using to iterate over tables [here](https://github.com/rmpestano/database-rider/blob/cfe8b3c3b57ca07267c6af9ad98b550b29daa2b2/rider-core/src/perf/java/com/github/database/rider/RiderDataSetBenchmark.java#L104-L124) but also would work with values too.

By the way I've created a [simple benchmark](https://github.com/rmpestano/database-rider/blob/master/rider-core/src/perf/java/com/github/database/rider/RiderDataSetBenchmark.java) to measure dataset creation via file (xml,json and yml) compared to dataset builder (in-memory) and the builder is like 10x faster than the fast file loader (JsonDataSet), here are the results:

```
[com.github.database.rider.RiderDataSetBenchmark.createDataSetsWithDataSetBuilder-jmh-worker-2] INFO com.github.database.rider.RiderDataSetBenchmark - Number of programmatic datasets created: 102794

[com.github.database.rider.RiderDataSetBenchmark.createDataSetsFromJSONFiles-jmh-worker-1] INFO com.github.database.rider.RiderDataSetBenchmark - Number of JSON datasets created: 10713

[com.github.database.rider.RiderDataSetBenchmark.createDataSetsFromXMLFiles-jmh-worker-2] INFO com.github.database.rider.RiderDataSetBenchmark - Number of XML datasets created: 7093

[com.github.database.rider.RiderDataSetBenchmark.createDataSetsFromYMLFiles-jmh-worker-4] INFO com.github.database.rider.RiderDataSetBenchmark - Number of YML datasets created: 3918
```

Following is the output of the benchmark in ops per sec (greater is better)
```
Benchmark                                                Mode  Cnt      Score      Error  Units
RiderDataSetBenchmark.createDataSetsFromJSONFiles       thrpt   15   6033,232 ± 1882,163  ops/s
RiderDataSetBenchmark.createDataSetsFromXMLFiles        thrpt   15   3877,318 ±  983,465  ops/s
RiderDataSetBenchmark.createDataSetsFromYMLFiles        thrpt   15   2118,727 ±  774,076  ops/s
RiderDataSetBenchmark.createDataSetsWithDataSetBuilder  thrpt   15  50567,545 ± 6668,360  ops/s

```

**Comment by rmpestano** [comment]
Hey folks, anything else we should consider before merging the PR?

I would like to thank you for all the ideia and your time discussing the DataSetBuilder API and reviewing the PR in details.

On the first moment I don't even consider It would be a good ideia to add such builder but now, thanks to you all, I'm not just certain but am sure that the DataSetBuilder can bring a lot more users to our API, so again thank you very much!

**Comment by rmpestano** [comment]
You can try new dataset builder API with dbrider v.1.5.4-SNAPSHOT by [enabling snapshots on your pom.xml](https://github.com/database-rider/database-rider#snapshots).

I plan to release 1.6.0 over next week.

**Comment by raderio** [comment]
Is there a way to specify that a table should be empty?
Something like
```
builder.table("user").empty()
```
instead of
```
<user />
```

**Comment by rmpestano** [comment]
Hi @raderio,

Just added support for empty tables, can you try with latest snapshot (you may need to run maven with `-U' flag).

Now this should work:

```
 IDataSet dataSet = builder.
                table("USER")
                .build();
```

or 

```
 builder.table("USER")
                .table("TWEET")
                .row()
                    .column("ID", "abcdef12345")
                    .column("CONTENT", "dbunit rules!")
                    .column("DATE", "[DAY,NOW]")
                .table("FOLLOWER")
...
...
```

**Comment by raderio** [comment]
Hi @rmpestano 
Thank you, for the improvement. In my opinion it will be better to call explicitly `empty` method.
What do you think about it?

**Comment by rmpestano** [comment]
Hi @raderio, any special reason for making it explicity? I think It would polute and make noise to the API for a very specific case.

**Comment by raderio** [comment]
It looks more clear and is clear that my intention is to have an empty table and not that I forgot to add some rows.
```
builder.table("USER").empty()
                .table("TWEET")
                .row()
                    .column("ID", "abcdef12345")
                    .column("CONTENT", "dbunit rules!")
                    .column("DATE", "[DAY,NOW]")
                .table("FOLLOWER").empty()
```

**Comment by rmpestano** [comment]
I see, but I think in both cases you'll have a failing test if you forget to add a row on your dataset. In the current approach your resulting dataset will have an empty table and on the proposed way you'll just don't have the table at all on your resulting dataset which I think will be harder to spot the root cause of the failing test. example:

**Comment by raderio** [comment]
Well, test can fail, but it also could not if the code has bugs, so if code will not insert some rows because of bugs, and in test I forgot to add rows, test will pass.

**Comment by rmpestano** [comment]
IMO it is pretty clear that code below will generate an empty dataset for user table: 

```
builder.table("USER").build();
```

I think you should extend DataSetBuilder in order to add the empty method.
